### PR TITLE
WIP Compositor

### LINKED ISSUE #40: Remove the kernel logger interface from userspace font rendering
The Logger trait and color attributes were used to multiplex kernel logging. They got copied to userspace when shell and clock were evicted from the kernel, but they could be hugely simplified now.

### PR DESCRIPTION
**Pull request by roblabla**
- Implement a compositor, named vi.
- The clock uses it.
- It's broken.

Fixes #40

### CODE REVIEW

**Comment by roblabla** [comment]
So, this works. It's slow though. Very, very slow.

**Comment by Orycterope** [comment]
Closes #40

File: libuser/src/terminal.rs [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+
+/// The font we choose to render in
+static FONT:  &'static [u8] = include_bytes!("../fonts/Monaco.ttf");
+
+/// The size we choose to render in
+const FONT_SIZE: u32 = 10;
+
+/// Window creation requested size.
+pub enum WindowSize {
+    Fullscreen, FontLines(i32, bool), Manual(i32, i32, u32, u32)
```
Comment (Orycterope) [inline libuser/src/terminal.rs]: What is this bool in FontLines ? How can we have a negative amount of lines ? 

Please document it. 
Also specify that FontLines takes the whole width of the screen (as I'm guessing).

File: libuser/src/terminal.rs [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+            .expect("Failed parsing provided font");
+
+        let my_ascent        =  my_font.max_ascent(FONT_SIZE).unwrap() as usize;
+        let my_descent       = -my_font.max_descent(FONT_SIZE).unwrap() as usize;
+        let my_advance_width =  my_font.max_advance_width(FONT_SIZE).unwrap() as usize;
+
+        let my_linespace = my_descent + my_ascent;
+
+        let framebuffer = match size {
+            WindowSize::Fullscreen => Window::new(0, 0, 1280, 800)?,
```
Comment (Orycterope) [inline libuser/src/terminal.rs]: add a todo
```rust
            // TODO: Terminal - Get fullscreen window size from vi
            // BODY: Terminal defines a fullscreen size as 1280 * 800, but should really
            // BODY: get it from vi instead. 
            WindowSize::Fullscreen => Window::new(0, 0, 1280, 800)?,
```

File: vi/src/main.rs [hunk 3]
```diff
@@ -1,59 +1,190 @@
...
+    width: u32,
+    height: u32,
+    mem: MappedSharedMemory
+}
+
+impl Buffer {
+    fn get_real_bounds(&self) -> (u32, u32, u32, u32) {
+        let dtop = max(self.top, 0) as u32;
+        let dleft = max(self.left, 0) as u32;
+        let dwidth = min(self.width as i32, self.left + (self.width as i32)) as u32;
```
[the remark this answers is not in the document]
Comment (roblabla) [reply vi/src/main.rs]: hmmm, shit, I'm going to need an if here.

File: vi/src/vbe.rs [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    /// Computes the offset in the framebuffer from x and y
+    ///
+    /// # Panics
+    ///
+    /// Panics if coords are invalid
+    #[inline]
+    pub fn write_px_at(&mut self, x: usize, y: usize, color: &VBEColor) {
+        let offset = self.get_px_offset(x, y);
+        if offset == 4096000 {
+            syscalls::output_debug_string(&format!("What the fuck: {} {}", x, y));
```
[the remark this answers is not in the document]
Comment (roblabla) [reply vi/src/vbe.rs]: welp, old debug stuff

File: vi/src/main.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Orycterope) [reply vi/src/main.rs]: I think you can do it with a 
```rust
max(0, min(...));
```

[the remark this answers is not in the document]
Comment (roblabla) [reply vi/src/main.rs]: Function was rewritten to have better semantics.

[the remark this answers is not in the document]
Comment (roblabla) [reply vi/src/main.rs]: Function was rewritten to have better semantics.

**Comment by roblabla** [comment]
@Orycterope Fixed everything. It's ready for review again.

File: libuser/src/window.rs [hunk 5]
```diff
@@ -107,22 +119,16 @@ impl Window {
...
         let fb = self.get_fb();
-        for i in fb.iter_mut() { *i = 0x00; }
-    }
-}
-
-/// Some colors for the vbe
-impl Color {
-    pub fn rgb(r: u8, g: u8, b: u8) -> Color {
-        Color {r, g, b }
+        for i in fb.iter_mut() { *i = Color::rgb(0, 0, 0); }
```
Comment (Orycterope) [inline libuser/src/window.rs]: Will this be optimized properly ? Or will this create a Color for every iteration ?

File: vi/src/vbe.rs [hunk 6]
```diff
@@ -52,6 +52,10 @@ impl Framebuffer {
         Ok(fb)
     }
 
+    pub fn new_buffer(buf: &'a mut [VBEColor], width: usize, height: usize, bpp: usize) -> Framebuffer<'a> {
```
Comment (Orycterope) [inline vi/src/vbe.rs]: documentation

File: libuser/src/window.rs (hunk 5 above)
Comment (Orycterope) [reply libuser/src/window.rs]: Note: Diff messed up, it's part of the clear screen function

Comment (roblabla) [reply libuser/src/window.rs]: yes

**Pull request merged**